File: extractor-appender/appender-sql/src/main/java/ru/mydesignstudio/database/metadata/extractor/database/type/TypeExtractor.java
package ru.mydesignstudio.database.metadata.extractor.database.type;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.database.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.database.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extract.result.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.extract.result.TypeModel;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

@Slf4j
@Component
public class TypeExtractor implements TableMetadataAppender {
  @StringResource("classpath:sql/extract_type.sql")
  private String typeQuery;

  @Autowired
  private ExtractHelper helper;

  @Override
  public void append(TableMetadata metadata, String schemaName, String tableName) {
    log.info("Extracting types for table {} in schema {}", tableName, schemaName);

    metadata.setTypes(helper.extract(typeQuery, new Object[]{tableName, schemaName}, TypeModel.class));

    log.debug("Extracted {} types", metadata.getTypes().size());
  }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/TypeExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.TypeModel

@Component
class TypeExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<TypeModel> {
        val query = queryProvider.provide("classpath:sql/extract_type.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
            TypeModel(
                    owner = getString(rs, "owner"),
                    objectName = getString(rs, "object_name"),
                    objectType = getString(rs, "object_type"),
                    comments = getString(rs, "comments")
            )
        }
    }
}